contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -0,0 +1,443 @@
+            order,
+            makerSignature,
+            nullSignature,
+            msg.value.safeDowncastToUint128(),
+            WethOptions.WrapEth
+        );
+    }
+
+    /// @dev Fully fill an OTC order. "Meta-transaction" variant,
+    ///      requires order to be signed by both maker and taker.
+    /// @param order The OTC order.
+    /// @param makerSignature The order signature from the maker.
+    /// @param takerSignature The order signature from the taker.
+    /// @param unwrapWeth Whether or not to unwrap bought WETH into ETH
+    ///        before transferring it to the taker. Should be set to false
+    ///        if the maker token is not WETH.
+    /// @return takerTokenFilledAmount How much taker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillTakerSignedOtcOrder(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        LibSignature.Signature memory takerSignature,
+        bool unwrapWeth
+    )
+        public
+        override
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        return _fillOtcOrderPrivate(
+            order,
+            makerSignature,
+            takerSignature,
+            order.takerAmount,
+            unwrapWeth ? WethOptions.UnwrapWeth : WethOptions.LeaveAsWeth
+        );
+    }
+
+    /// @dev Fill an OTC order. Private variant.
+    /// @param order The OTC order.
+    /// @param makerSignature The order signature from the maker.
+    /// @param takerSignature The order signature from the taker.
+    ///        Ignored if msg.sender == order.taker.
+    /// @param takerTokenFillAmount Maximum taker token amount to
+    ///        fill this order with.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _fillOtcOrderPrivate(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        LibSignature.Signature memory takerSignature,
+        uint128 takerTokenFillAmount,
+        WethOptions wethOptions
+    )
+        private
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        LibNativeOrder.OtcOrderInfo memory orderInfo = getOtcOrderInfo(order);
+
+        // Must be fillable.
+        if (orderInfo.status != LibNativeOrder.OrderStatus.FILLABLE) {
+            LibNativeOrdersRichErrors.OrderNotFillableError(
+                orderInfo.orderHash,
+                uint8(orderInfo.status)
+            ).rrevert();
+        }
+
+        {
+            LibNativeOrdersStorage.Storage storage stor =
+                LibNativeOrdersStorage.getStorage();
+
+            // Must be fillable by the tx.origin.
+            if (
+                order.txOrigin != tx.origin &&
+                !stor.originRegistry[order.txOrigin][tx.origin]
```

Comment:
How do we add data to the `originRegistry`? Is there a getter/setter?

Reply:
It's the same `originRegistry` as RFQ orders

Reply:
Thanks!